source/MaterialXShaderGen/ShaderGenerators/Glsl/BitangentGlsl.cpp

```diff
@@ -0,0 +1,90 @@
+#include <MaterialXShaderGen/ShaderGenerators/Glsl/BitangentGlsl.h>
+
+namespace MaterialX
+{
+
+SgImplementationPtr BitangentGlsl::creator()
+{
+    return std::make_shared<BitangentGlsl>();
+}
+
+void BitangentGlsl::createVariables(const SgNode& node, ShaderGenerator& /*shadergen*/, Shader& shader_)
+{
+    HwShader& shader = static_cast<HwShader&>(shader_);
+
+    shader.createAppData(DataType::VECTOR3, "i_bitangent");
+
+    const SgInput* spaceInput = node.getInput(SPACE);
+    string space = spaceInput ? spaceInput->value->getValueString() : EMPTY_STRING;
+    if (space == WORLD)
+    {
+        shader.createUniform(HwShader::GLOBAL_SCOPE, DataType::MATRIX4, "u_normalMatrix");
+        shader.createVertexData(DataType::VECTOR3, "bitangentWorld");
```

Comment:
Exactly. The block named "VertexData" is used for vertex -> pixel shader transfer (varyings in the deprecated nomenclature). So the vertex shader will declare this as an output block and the pixel shader will declare it as an input block.

The matrix is added as a uniform in the "global scope" block so it can be used by the vertex shader to do the transformation. However global uniforms are also emitted to pixel shader, which is not really needed in this case. So it could be a better design to separate this per stange instead (as you suggested below).

Reply:
Thanks for the explanation, I'm very old school and still used to the term "connectors" :)